### ISSUE TITLE
Excluding weight decay not working properly on most LMs

### ISSUE THREAD

**Issue by wwt17** [huggingface/transformers#42754]
The problem is consistent across all three files:
https://github.com/huggingface/transformers/blob/471d7ce9abbb3bc1b3bab673367378f9dbc3caac/examples/pytorch/language-modeling/run_clm_no_trainer.py#L518-L528
https://github.com/huggingface/transformers/blob/471d7ce9abbb3bc1b3bab673367378f9dbc3caac/examples/pytorch/language-modeling/run_fim_no_trainer.py#L710-L720
https://github.com/huggingface/transformers/blob/471d7ce9abbb3bc1b3bab673367378f9dbc3caac/examples/pytorch/language-modeling/run_mlm_no_trainer.py#L544-L554
Many LMs use patterns like `[r"layernorm", r"rmsnorm", r"(?:^|\.|_)norm(?:$|\.|_)", r"(?:^|\.|_)ln(?:$|\.|_)"]`, so the simple `no_decay` patterns here won't exclude many normalization layers. I also suggest excluding embeddings and `r"(?:^|\.)wpe(?:$|\.)"` to exclude input/output/positional embeddings.
Could consider using/modifying from the code here:
https://github.com/huggingface/transformers/blob/471d7ce9abbb3bc1b3bab673367378f9dbc3caac/src/transformers/trainer.py#L1213-L1227

**Comment by casinca** [comment]
Hi, I've opened a PR for these files and re-used the logic from `trainer.py` as you mentioned but for the embeddings, we'll see what the HF team wants to do.

**Comment by azurearchitect786** [comment]
If everything's clear, post a comment on the issue to indicate your interest to work on it.

**Comment by wwt17** [comment]
Thank you! Let's wait and see!

**Comment by casinca** [comment]
Sorry for the delay, I forgot about the PR, just wanted to let you know in case, it was pushed yesterday https://github.com/huggingface/transformers/pull/42769
